Lib/test/test_capi/test_unicode.py

```diff
@@ -1297,6 +1297,60 @@ def test_comparewithasciistring(self):
         # CRASHES comparewithasciistring([], b'abc')
         # CRASHES comparewithasciistring(NULL, b'abc')
 
+    @support.cpython_only
+    @unittest.skipIf(_testcapi is None, 'need _testcapi module')
+    def test_equaltoutf8(self):
+        """Test PyUnicode_EqualToUTF8()"""
+        from _testcapi import unicode_equaltoutf8 as equaltoutf8
+        from _testcapi import unicode_asutf8andsize as asutf8andsize
+
+        strings = [
+            'abc', '\xa1\xa2\xa3', '\u4f60\u597d\u4e16',
+            '\U0001f600\U0001f601\U0001f602',
+            '\U0010ffff',
+        ]
+        for s in strings:
+            # Call PyUnicode_AsUTF8AndSize() which creates the UTF-8
+            # encoded string cached in the Unicode object.
+            asutf8andsize(s, 0)
+            b = s.encode()
+            self.assertEqual(equaltoutf8(s, b), 1)  # Use the UTF-8 cache.
+            s2 = b.decode()  # New Unicode object without the UTF-8 cache.
+            self.assertEqual(equaltoutf8(s2, b), 1)
+            self.assertEqual(equaltoutf8(s + 'x', b + b'x'), 1)
+            self.assertEqual(equaltoutf8(s + 'x', b + b'y'), 0)
+            self.assertEqual(equaltoutf8(s + '\0', b + b'\0'), 0)
+            self.assertEqual(equaltoutf8(s2, b + b'x'), 0)
+            self.assertEqual(equaltoutf8(s2, b[:-1]), 0)
+            self.assertEqual(equaltoutf8(s2, b[:-1] + b'x'), 0)
+
+        # embedded null chars/bytes
+        self.assertEqual(equaltoutf8('abc', b'abc\0def\0'), 1)
```

Comment:
This is... annoying to say the least. Can `PyUnicode_EqualToUTF8` instead take a third argument `Py_ssize_t len` argument?

Reply:
It is inconvenient and errorprone to specify a size of literal strings.

Reply:
that's why I suggest a pair of APIs, one primarily intended for literals (I wish C could enforce that), one with the size.

Reply:
Adding PyUnicode_EqualToUTF8Size() (or PyUnicode_EqualToUTF8WithSize() if you love old habits and long names) would make sense.

 I like PyUnicode_EqualToUTF8(). For me truncating the second argument at the first NUL byte is the classic way to handle strings in C. All "str" functions of the libc API are based on this format, only "mem" functions take a size/length parameters.

Reply:
> For me truncating the second argument at the first NUL byte is the classic way to handle strings in C

It's certainly classic, but it's a bad kind of classic :-)